sql/schemachanger: fallback to legacy schema changer when zone configs exist

fqazi opened the pull request:
We enabled a number of operations inside the declarative schema changer by default, these operations do not
properly support updating the zone configuration. As a stop-gap solution, we are going to revert to the legacy
schema changer if zone configs exist on tables for ADD/DROP COLUMN, CREATE DROP/INDEX, ALTER PRIMARY KEY.

Release justification: low risk change that uses legacy schema changer in scenarios where we know the declarative one may behave incorrectly.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86326)
<!-- Reviewable:end -->

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go:
```diff
@@ -39,6 +39,7 @@ func alterTableDropColumn(
 		return
 	}
 	checkColumnNotInaccessible(col, n)
+	fallBackIfZoneConfigExists(b, n, tbl.TableID)
```
In the PR description, you mentioned that we will do this fall-back-if-zone-config-exists for ADD/DROP COLUMN, CREATE/DROP INDEX, and ALTER PRIMARY KEY. I only see you added this to `DROP COLUMN`, did you forget to add this to the other stmt, or did I miss something?

Xiang-Gu replied:
nvm, I was looking at only the first commit. I saw it in the second commit

postamar commented:
Isn't this also missing index zone configs and index partition zone configs?

fqazi commented:
> Isn't this also missing index zone configs and index partition zone configs?
So, that's because we store the index and partition zone config inside the table descriptor zone config. This approach pretty much says if we see *any* table zone config then these operations will be disallowed. I think for a fuller decomposition we definitely need those elements though and ones for partitions. This is a fairly blunt approach

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86326)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86326#-N9h32uw0Ti7_9RT11FN-r1-42:-N9lT3Ri9SitE4Nju3wT:b-h3owkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3403ea690211377105db2d7deed343555da8946/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L42)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

nvm, I was looking at only the first commit. I saw it in the second commit
</blockquote></details>

Let me move that file between commits, I accidentally pushed one of them in the first commit and clobbered some code


<!-- Sent from Reviewable.io -->

postamar commented:
> So, that's because we store the index and partition zone config inside the table descriptor zone config. This approach pretty much says if we see any table zone config then these operations will be disallowed. I think for a fuller decomposition we definitely need those elements though and ones for partitions. This is a fairly blunt approach

Ah, thanks, this makes sense. This blunt approach is fine for the time being. I'll review in more depth ASAP.

fqazi committed:
sql/schemachanger: add interfaces needed to read zone configs

Previously, we the declarative schema changer had no
way of reading zone configs, and generating elements
for them. This was inadequate because we needed a
way to generate these elements, so that certain operations
could be blocked. To address this, this patch adds very basic
boiler plate elements for zone configs at the table level,
which are sufficient for throwing not implemented errors.
Release justification: Only adds interfaces to help fix a bug
Release note: None

postamar commented on pkg/sql/schemachanger/scdeps/sctestdeps/database_state.go:
```diff
@@ -136,6 +139,32 @@ ORDER BY id`)
+		defer zoneCfgRow.Close()
+		once := false
+		for zoneCfgRow.Next() {
+			require.Equalf(t, false, once, "multiple zone config entries for descriptor")
```
nit: require.Falsef

postamar commented on pkg/sql/schemachanger/scdecomp/testdata/table:
```diff
@@ -11,6 +11,7 @@ COMMENT ON TABLE tbl IS 'tbl is good table';
 COMMENT ON INDEX tbl@tbl_pkey IS 'tbl_pkey is a primary key';
 COMMENT ON COLUMN tbl.id IS 'id is a identifier';
 COMMENT ON CONSTRAINT myfk ON tbl IS 'must have a parent';
+ALTER TABLE tbl CONFIGURE ZONE USING gc.ttlseconds=10;
```
nit: can you add a case (or replace this one) where you add a zone config to an index instead of the table itself?

postamar commented on pkg/sql/schemachanger/scdecomp/dependencies.go:
```diff
@@ -44,6 +46,13 @@ type CommentGetter interface {
 
+// ZoneConfigReader supports reading raw zone config information
+// from storage.
+type ZoneConfigReader interface {
```
uber nit: ZoneConfigGetter? for the sake of uniformity with CommentGetter.

postamar commented on pkg/sql/schemachanger/scdecomp/decomp.go:
```diff
@@ -318,6 +322,21 @@ func (w *walkCtx) walkRelation(tbl catalog.TableDescriptor) {
+	// Add a zone config element which is a stop gap to allow us to block
+	// operations on tables. To minimize RTT impact limit this to only
+	// tables.
+	if tbl.IsTable() {
```
According to the element model definition that you added, zone configs can also exist for views and sequences. This line is inconsistent with that, please address.

postamar commented on pkg/sql/schemachanger/scbuild/testdata/unimplemented_zone_cfg:
```diff
@@ -0,0 +1,31 @@
+ALTER TABLE defaultdb.foo CONFIGURE ZONE USING gc.ttlseconds=10;
+CREATE INDEX idx ON defaultdb.foo(k, l);
+----
+
```
nit: can you also add a CREATE INDEX test case?

postamar commented on pkg/sql/schemachanger/scpb/uml/table.puml:
```diff
@@ -101,6 +101,10 @@ object RowLevelTTL
+object TableZoneConfig
+
+TableZoneConfig :  TableID
+
```
nit: these uml diagrams an `elements_generated.go` changes belong in the other commit

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_index.go:
```diff
@@ -116,6 +116,9 @@ func dropAnIndex(
 	}
+	// Disable dropping indexes if the table could have zone configuration
+	// settings.
+	fallBackIfZoneConfigExists(b, nil, sie.TableID)
```
nit: this call should take place far earlier

postamar commented on pkg/sql/schemachanger/scdeps/sctestdeps/config.go:
```diff
@@ -79,6 +79,13 @@ func WithSessionData(sessionData sessiondata.SessionData) Option {
+	return optionFunc(func(state *TestState) {
+		state.zoneConfigs = zoneConfigs
+	})
+}
```
nit: doesn't this belong in the first commit?

postamar commented on pkg/sql/schemachanger/scdeps/sctestdeps/test_deps.go:
```diff
@@ -1187,3 +1187,13 @@ func (s *TestState) ResolveFunctionByOID(
+// GetZoneConfig implements scexec.Dependencies.
+func (s *TestState) GetZoneConfig(ctx context.Context, id descpb.ID) (*zonepb.ZoneConfig, error) {
+	return s.zoneConfigs[id], nil
+}
```
nit: don't these belong in the first commit?

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -102,7 +102,7 @@ func CreateIndex(b BuildCtx, n *tree.CreateIndex) {
 		}
 	})
-
+	fallBackIfZoneConfigExists(b, n, index.TableID)
```
nit: please call this as early as possible

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go:
```diff
@@ -103,6 +103,10 @@ func alterTableAddColumn(
 	}
+	// We don't support handling zone config related properties for tables, so
+	// throw an unsupported
+	fallBackIfZoneConfigExists(b, d, tbl.TableID)
```
nit: please call this earlier

postamar reviewed (commented):
This is close, just one non-nit concern I'd like addressed.

fqazi committed:
sql/schemachanger: disable operations that require zone config support

Previously, we incorrectly allowed operations that require
mutations to the zone config to be allowed inside the
declarative schema changer. This was inadequate because
the declarative schema changer did not correctly mutate
the zone config which could lead to invalid schemas. To
address this, this patch will throw not implemented errors,
when zone config objects are detected.

Release justification: Low risk and fixes a bug inside the code
that isn't exposed to users yet.
Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86326)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 328 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86326#-N9h32uw0Ti7_9RT11FP-r3-328:-N9mo4w-4SLME_scwwJs:b-b7fk7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/476f20c36839a4d6e9fe9ca0bef3a8311c807213/pkg/sql/schemachanger/scdecomp/decomp.go#L328)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

According to the element model definition that you added, zone configs can also exist for views and sequences. This line is inconsistent with that, please address.
</blockquote></details>

Good point I'm gonna remove it for views/sequences for now in the element model. Trying to avoid unnecessary RTT time
___
*[`pkg/sql/schemachanger/scdecomp/dependencies.go` line 51 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86326#-N9h32uw0Ti7_9RT11FQ-r3-51:-N9moGJXDMakkUU0a3Pe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/476f20c36839a4d6e9fe9ca0bef3a8311c807213/pkg/sql/schemachanger/scdecomp/dependencies.go#L51)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

uber nit: ZoneConfigGetter? for the sake of uniformity with CommentGetter.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

postamar reviewed (approved):
LGTM

fqazi commented:
@postamar TFTR

bors r+

fqazi commented:
bors r+
